libp2p-xtra/src/listener.rs

```diff
@@ -0,0 +1,120 @@
+use crate::ConnectionStats;
+use crate::Endpoint;
+use crate::GetConnectionStats;
+use crate::ListenOn;
+use anyhow::anyhow;
+use anyhow::Result;
+use async_trait::async_trait;
+use futures::TryFutureExt;
+use libp2p_core::Multiaddr;
+use std::time::Duration;
+use tokio_tasks::Tasks;
+use xtra::Address;
+use xtra::Disconnected;
+use xtra_productivity::xtra_productivity;
+use xtras::SendInterval;
+
+/// How often we check whether we are still connected
+pub const CHECK_INTERVAL: Duration = Duration::from_secs(5);
+
+/// xtra actor that taker care of listening for incoming connections to the Endpoint.
+///
+/// Periodically polls Endpoint to check whether connection is still active.
+/// Should be used in conjunction with supervisor for continuous and resilient listening.
+pub struct Actor {
+    tasks: Tasks,
+    endpoint: Address<Endpoint>,
+    listen_address: Multiaddr,
+    /// Contains the reason we are stopping.
+    stop_reason: Option<Error>,
+}
+
+impl Actor {
+    pub fn new(endpoint: Address<Endpoint>, listen_address: Multiaddr) -> Self {
+        Self {
+            tasks: Tasks::default(),
+            endpoint,
+            listen_address,
+            stop_reason: None,
+        }
+    }
+
+    async fn check_connection_active_in_endpoint(&self) -> Result<()> {
+        let ConnectionStats {
+            listen_addresses, ..
+        } = self
+            .endpoint
+            .send(GetConnectionStats)
+            .await
+            .map_err(|_| Error::NoEndpoint)?;
+
+        listen_addresses
+            .contains(&self.listen_address)
+            .then(|| ())
+            .ok_or_else(|| anyhow!("Listener address not active in Endpoint"))
+    }
+}
+
+#[async_trait]
+impl xtra::Actor for Actor {
+    type Stop = Error;
+
+    async fn started(&mut self, ctx: &mut xtra::Context<Self>) {
+        let this = ctx.address().expect("we are alive");
+
+        let endpoint = self.endpoint.clone();
+        let listen_address = self.listen_address.clone();
+
+        {
+            let this = this.clone();
+            self.tasks.add_fallible(
+                endpoint
+                    .send(ListenOn(listen_address))
+                    .map_err(|e| Error::Failed { source: anyhow!(e) }),
+                move |e| async move {
+                    tracing::error!("Cannot listen on the endpoint");
+                    let _: Result<(), Disconnected> = this.send(e).await;
+                },
+            );
+        };
+
+        // Only start checking the connection after it had enough time to be established
+        tokio::time::sleep(CHECK_INTERVAL).await;
+        self.tasks
+            .add(this.send_interval(CHECK_INTERVAL, || CheckConnection));
```

Comment:
❓ What about this?

```suggestion
        if let Err(e) = endpoint.send(ListenOn(listen_address)).await {
            self.stop_reason = ErrorFailed { source: anyhow!(e) };
            ctx.stop()
        };
    
        self.tasks
            .add(this.send_interval(CHECK_INTERVAL, || CheckConnection));
```

I think we can stop immediately if we can't listen on the endpoint. That way we don't have to introduce the `sleep` and the logic is simpler.

Reply:
we still need `sleep` because the Endpoint takes a while for it to register that, it's not immediately available :(

I reworked the other part, it is simpler indeed!

Reply:
Shame about the sleep! I wonder if sending `ListenOn` should only return after the `Endpoint` is ready to listen. That may not be possible though.

Nice otherwise!

Reply:
Another thing that I could think of would be setting it up immediately in the hashset in `Endpoint`. This was indeed the original behaviour, but somehow we changed it to be more correct. 
In other words, we could redefine from what it means to be present in the hashset to: "listening or listening requested". It would still be removed if the request failed, so I guess it's more ergonomic for users, especially when polling is the way.

Thoughts?